### PR TITLE
[14.0] Shopfloor, checkout: Allow to scan products

### PR DESCRIPTION
**Pull request by mmequignon**
based on https://github.com/OCA/wms/pull/404 to avoid conflicts

### CODE REVIEW

File: shopfloor/services/checkout.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def _response_for_change_packaging(self, picking, package, packaging_list):
         )
 
     def scan_document(self, barcode):
-        """Scan a package, a stock.picking or a location
+        """Scan a package, a product, a stock.picking or a location
```
Comment (sebalix) [inline shopfloor/services/checkout.py]: We could even reword `stock.picking` to `transfer` as it's a docstring.

File: shopfloor/services/checkout.py [hunk 2]
```diff
@@ -216,6 +220,33 @@ def scan_document(self, barcode):
...
+            line_domain = []
+            if not product:
+                # Then, try to find a packaging matching the barcode
+                packaging = search.packaging_from_scan(barcode)
+                if packaging:
+                    # And retrieve its product
+                    product = packaging.product_id
+                    # The picking should have a move line for the product
+                    # where qty > packaging.qty
+                    line_domain.append(("product_uom_qty", ">", packaging.qty))
```
Comment (sebalix) [inline shopfloor/services/checkout.py]: Not sure to understand this criteria?

Comment (mmequignon) [reply shopfloor/services/checkout.py]: Well, as I understand it, it doesn't makes sense to select a move line which have less qty than the selected packaging.

File: shopfloor/actions/search.py [hunk 3]
```diff
@@ -58,6 +58,14 @@ def lot_from_scan(self, barcode, products=None, limit=1):
             domain.append(("product_id", "in", products.ids))
         return model.search(domain, limit=limit)
 
+    def packaging_from_scan(self, barcode):
```
Comment (TDu) [inline shopfloor/actions/search.py]: Looks like the same function than the one below ?

File: shopfloor/models/shopfloor_menu.py [hunk 4]
```diff
@@ -11,6 +11,21 @@
 * in both cases, if the picking has no carrier the operation fails.",
 """
 
+UNLOAD_PACK_AT_DEST_HELP = """
+With this option, the lines you processed by putting on a package during the
```
Comment (TDu) [inline shopfloor/models/shopfloor_menu.py]:
```suggestion
With this option, the lines you process by putting them in a package during the
```

File: shopfloor/models/shopfloor_menu.py [hunk 5]
```diff
@@ -71,6 +114,63 @@ class ShopfloorMenu(models.Model):
         compute="_compute_allow_force_reservation_is_possible"
     )
 
+    @api.onchange("unload_package_at_destination")
+    def _onchange_unload_package_at_destination(self):
+        # Uncheck pick_pack_same_time when unload_package_at_destination is set to True
+        # Ensure that multiple_move_single_pack is False when
+        # unload_package_at_destination is checked out
+        for record in self:
+            if self.unload_package_at_destination:
```
Comment (TDu) [inline shopfloor/models/shopfloor_menu.py]:
```suggestion
            if record.unload_package_at_destination:
```
But I don't think onchange can receive a recordset. No need of the loop on self.

**Review by ivantodorovich — APPROVED** [review]

File: shopfloor/models/shopfloor_menu.py (hunk 5 above)
Comment (mmequignon) [reply shopfloor/models/shopfloor_menu.py]: This one has been introduced in another pr https://github.com/OCA/wms/pull/398

File: shopfloor/services/checkout.py [hunk 6]
```diff
@@ -216,6 +220,34 @@ def scan_document(self, barcode):
...
+            if not product:
+                # Then, try to find a packaging matching the barcode
+                packaging = search.packaging_from_scan(barcode)
+                if packaging:
+                    # And retrieve its product
+                    product = packaging.product_id
+                    # The picking should have a move line for the product
+                    # where qty > packaging.qty, since it doesn't makes sense
+                    # to select a move line which have less qty than the packaging
+                    line_domain.append(("product_uom_qty", ">", packaging.qty))
```
Comment (jbaudoux) [inline shopfloor/services/checkout.py]:
```suggestion
                    line_domain.append(("product_uom_qty", ">=", packaging.qty))
```

File: shopfloor/services/checkout.py [hunk 7]
```diff
@@ -216,6 +220,34 @@ def scan_document(self, barcode):
...
+                line_domain.extend(
+                    [
+                        ("product_id", "=", product.id),
+                        ("state", "not in", ("cancel", "done")),
+                        ("package_id", "=", False),
+                    ]
+                )
+                lines = self.env["stock.move.line"].search(line_domain)
+                picking = self.env["stock.picking"].search(
+                    [("id", "in", lines.picking_id.ids)],
```
Comment (jbaudoux) [inline shopfloor/services/checkout.py]: and picking_type_id in self.picking_types

**Review by jbaudoux — CHANGES_REQUESTED** [review]

**Review by sebalix — APPROVED** [review]

**Pull request merged**